### ISSUE TITLE
Logical AND (&&) result incorrectly type checked in JSX files

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#19491]
**TypeScript Version:**  2.4.1 and 2.6.0-rc

**Code**

```ts
{
    const boolean = false as boolean;

    // Expected error, got error
    // Type 'false' is not assignable to type 'number'.
    const x: number = boolean && 1;
}
```

The above example correctly gives a type checking error. However, the same code in a JSX file (`.tsx`) does not give an error:

``` tsx
{
    const boolean = false as boolean;

    // Expected error, but got no error!
    const x: number = boolean && 1;
}
```

**Comment by andy-ms** [comment]
Works for me -- is it possible these are two different projects with different settings for `--strictNullChecks`?

**Comment by RyanCavanaugh** [comment]
This is just because one of your files got into a tsconfig file that set `strictNullChecks`

**Comment by OliverJAsh** [comment]
Aha, `strictNullChecks` is related. I was wrong about this only appearing in JSX files, apologies.

If I enable `strictNullChecks`, the expected error appears.

However, I don't understand why the error doesn't appear when `strictNullChecks` is disabled?

**Comment by andy-ms** [comment]
It was a bug from before `--strictNullChecks` and fixing it would break backwards compatibility. So we only fixed it in the `--strictNullChecks` case.

**Comment by OliverJAsh** [comment]
Aha, thanks @andy-ms! Will these things be ironed out eventually in the next major version, along with other breaking chances?

**Comment by andy-ms** [comment]
As far as I know there aren't any plans to make breaking changes.
